src/vendorcode/fsp/qemu/src/lib.rs

```diff
@@ -22,3 +23,59 @@ macro_rules! blob_macro {
 #[used]
 #[link_section = ".fspblob"]
 static FSP_BLOB: [u8; blob_macro!().len()] = *blob_macro!();
+
+pub fn get_fspm_upd() -> FSPM_UPD {
+    // TODO: This struct has to be aligned to 4.
+    // mut because we can't make the assumption FSP won't modify it.
+    FSPM_UPD {
+        FspUpdHeader: FSP_UPD_HEADER {
+            Signature: FSPM_UPD_SIGNATURE,
+            Revision: 2, // FSP 2.2
+            Reserved: [0u8; 23],
+        },
+        FspmArchUpd: FSPM_ARCH_UPD {
+            Revision: 2, // FSP 2.2
+            Reserved: [0u8; 3],
+            NvsBufferPtr: 0,        // non-volatile storage not available
+            StackBase: 0x20000000,  // TODO: I picked this at random
+            StackSize: 0x10000,     // TODO: I picked this at random
+            BootLoaderTolumSize: 0, // Don't reserve "top of low usable memory" for bootloader.
+            BootMode: BOOT_WITH_FULL_CONFIGURATION,
+            FspEventHandler: 0 as *mut FSP_EVENT_HANDLER, // optional
+            Reserved1: [0u8; 4],
+        },
+        FspmConfig: FSP_M_CONFIG {
+            SerialDebugPortAddress: 0x3f8,
+            SerialDebugPortType: 1,       // I/O
+            SerialDebugPortDevice: 3,     // External Device
+            SerialDebugPortStrideSize: 0, // 1
+            UnusedUpdSpace0: [0; 49],
+            ReservedFspmUpd: [0; 4],
+        },
+        UnusedUpdSpace1: [0u8; 2],
+        UpdTerminator: 0x55AA, // ???
+    }
+}
+
+pub fn get_fsps_upd() -> FSPS_UPD {
+    // TODO: This struct has to be aligned to 4.
```

Comment:
Titus looked and this should be OK because that is C, and Rust should assume C alignment and ordering rules apply?

Reply:
Would be happy to hear this in the case 😄  I just retained this comment from the original implementation. But if we think it's irrelevant, happy to remove it

Reply:
I think we don't need the TODO anymore. It's repr C and packed.

Reply:
Confirmed that this is the case: https://doc.rust-lang.org/reference/type-layout.html

"The alignment of the struct is the alignment of the most-aligned field in it."

So removed.